### PR TITLE
Xapic support

### PR DESCRIPTION
**Pull request by vinaychandra**
This is a fork simplified from the other [pull request](https://github.com/kzhao0986/x2apic-rs/pull/1)

### CODE REVIEW

**Comment by vinaychandra** [comment]
Although the logic is pretty much taken from the other PR, this doesn't work as expected atleast for me. Please validate before merging.

**Comment by kzhao0986** [comment]
Thanks, @vinaychandra for this! Even after fixing a pointer reading bug (my bug when I was sketching the design in the previous PR), I'm still hitting a double fault exception when I try to run things in xAPIC mode.

I'm a little strapped for time this week, but I'll try looking into this soon, especially since it looks like it'd be useful to many. I'm also starting to think the original code could use some restructuring to make adding this easier.

**Comment by michaelmelanson** [comment]
I'm coming back to this now finally. 😅 I'll close my PR and we can work together on landing this version.

File: src/lapic/lapic_msr.rs [hunk 1]
```diff
@@ -8,31 +8,37 @@ use x86_64::registers::model_specific::Msr;
...
     }
 
-    pub unsafe fn write(&mut self, mode: &LocalApicMode, value: u64) {
-        match mode {
-            LocalApicMode::XApic => { *(self.mmio_address() as *mut u64) = value },
-            LocalApicMode::X2Apic => self.msr.write(value)
+    pub unsafe fn write(&mut self, value: u64) {
+        match self {
+            LocalApicRegister::XapicOffset(offset) => {
+                *(offset as *mut u64) = value
```
Comment (Freax13) [inline src/lapic/lapic_msr.rs:39]: There are two subtle errors here:
1. `offset` is a reference to the field in the enum, so `offset as *mut u64` gets a pointer to this field, but really we wanted to cast the value of the field to a pointer. The correct version would be `*offset as *mut u64`
2. You should use `core::ptr::write_volatile` here to make sure the read isn't optimized away or reordered

The corrected code should look like this:
`core::ptr::write_volatile(*offset as *mut u64, value)`

File: src/lapic/lapic_msr.rs [hunk 2]
```diff
@@ -8,31 +8,37 @@ use x86_64::registers::model_specific::Msr;
...
     }
 
-    pub unsafe fn read(&self, mode: &LocalApicMode) -> u64 {
-        match mode {
-            LocalApicMode::XApic => *(self.mmio_address() as *const u64),
-            LocalApicMode::X2Apic => self.msr.read()
+    pub unsafe fn read(&self) -> u64 {
+        match self {
+            Self::XapicOffset(addr) => {
+                core::ptr::read_volatile(addr as *const u64)
```
Comment (Freax13) [inline src/lapic/lapic_msr.rs:30]: Same thing here: this should be `core::ptr::read_volatile(*addr as *const u64)`

**Comment by kwzhao** [comment]
Closing in favor of #7

**Pull request closed without merging**